pkg/server/singleprocess/service_pipeline.go

```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
 			}
 		}
 
-		run.Jobs = append(run.Jobs, &pb.Ref_Job{Id: job.Id})
+		pipelineRun.Jobs = append(pipelineRun.Jobs, &pb.Ref_Job{Id: job.Id})
 		stepJobs = append(stepJobs, &pb.QueueJobRequest{
 			Job: job,
 		})
 	}
 
-	// Get the graph for the steps so we can get the root. We enforce a
-	// single root so the root is always the first step.
-	stepGraph, err := serverptypes.PipelineGraph(pipeline)
-	if err != nil {
-		return nil, err
+	return stepJobs, pipelineRun, stepIds, nil
+}
+
+// pipelineGraphFull takes a pipeline, and optionally accepts an existing Graph
+// and parent step, and attempts to build a full graph for a given Pipeline including
+// any nested pipeline steps. It keeps track of visisted steps so that we don't
+// get stuck in a loop building the graph. This means step names must be unique
+// across pipelines.
+func (s *Service) pipelineGraphFull(
+	ctx context.Context,
+	log hclog.Logger,
+	g *graph.Graph,
+	parentStep string,
+	visistedNodes map[string]string,
+	nodeIdMap map[string]*pb.Ref_PipelineStep,
+	pipeline *pb.Pipeline,
+) (*graph.Graph, map[string]*pb.Ref_PipelineStep, error) {
+	var stepGraph *graph.Graph
+	if g != nil {
+		// We're handling an embedded pipeline graph
+		stepGraph = g
+	} else if stepGraph == nil {
+		stepGraph = &graph.Graph{}
 	}
```

Comment:
What is the point of this `else` if `stepGraph` is declared as nil on line 416 and remains nil unless the conditional is hit?

Reply:
The main point here is initializing `stepGraph` from nil into an empty `*graph.Graph{}` struct in the base case or first iteration of this func that I'm handling. I figured it might be worth being explicit assigning `stepGraph` when it's nil rather than an empty struct, but might not matter?

Reply:
```suggestion
	stepGraph := &graph.Graph{}
	if g != nil {
		// We're handling an embedded pipeline graph
		stepGraph = g
```
Yea, the way the code is written now, it's superfluous. I would declare it as the empty graph if that's the desired state if `g != nil`.

Reply:
Yeah makes sense! I'll update it 👍🏻